### PR TITLE
Optimize check for missing fields in deserialization

### PR DESCRIPTION
**Pull request by shanshin**

### CODE REVIEW

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt [hunk 1]
```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
...
+                    goldenMask = goldenMask or requiredBit
+                }
+                requiredBit = requiredBit shl 1
+            }
+
+            iconst(goldenMask)
+            dup()
+            load(maskVar, OPT_MASK_TYPE)
+            and(OPT_MASK_TYPE)
+            ificmpeq(allPresentsLabel)
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:303]: It looks like this solution does not work if we have optional fields. In this case, seen mask contains more one bits than the golden one, and therefore they're not equal.

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt [hunk 2]
```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
...
+
+            load(maskVar, OPT_MASK_TYPE)
+            iconst(goldenMask)
+            load(descVar, descType)
+            invokestatic(
+                "kotlinx/serialization/internal/$PLUGIN_UTILS_LIBRARY_CLASS", THROW_MISSING_FIELD_EXCEPTION_METHOD,
+                "(IILkotlinx/serialization/descriptors/$SERIAL_DESCRIPTOR_CLASS;)V", false
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:310]: again, `descType.descriptor`

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt [hunk 3]
```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
...
+        val goldenMaskArray = IntArray(maskSlotCount)
+
+        for (i in properties.serializableProperties.indices) {
+            if (!properties.serializableProperties[i].optional) {
+                val slotNumber = i / 32
+                val bitInSlot = i % 32
+                goldenMaskArray[slotNumber] = goldenMaskArray[slotNumber] or (1 shl bitInSlot)
+            }
+        }
+        return goldenMaskArray;
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:365]:
```suggestion
        return goldenMaskArray
```

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt [hunk 4]
```diff
@@ -261,12 +277,94 @@ class SerializableCodegenImpl(
...
+                iconst(goldenMask)
+                dup()
+                load(maskIndex, OPT_MASK_TYPE)
+                and(OPT_MASK_TYPE)
+                ificmpne(fieldsMissingLabel)
+            }
+            goTo(allPresentsLabel)
+            //prepare seen array
+            visitLabel(fieldsMissingLabel)
+            iconst(maskSlotCount)
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:328]: Maybe you'll find `InstructionAdapter.fillArray` function from JVMCodegenUtil.kt useful

**Review by sandwwraith — CHANGES_REQUESTED** [review]
Also, it seems your commit message refers to issues in the Kotlin repo, not kotlinx.serialization one. You can paste full links instead.

Notice that `SerializationPluginBytecodeListingTestGenerated` will likely fail after your changes due to changed bytecode shape. You can delete `testData/codegen/basic.txt` and re-run test to generate new bytecode

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt (hunk 1 above)
Comment (shanshin) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:303]: Seen mask can't contains more bits because first run "and" operation and only then comparing with golden mask: (seen and golden) == golden

Comment (sandwwraith) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/jvm/SerializableCodegenImpl.kt:303]: hm, indeed!

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/NamingConventions.kt [hunk 5]
```diff
@@ -68,6 +68,8 @@ object SerialEntityNames {
     const val SERIAL_DESCRIPTOR_CLASS_IMPL = "PluginGeneratedSerialDescriptor"
     const val SERIAL_DESCRIPTOR_FOR_ENUM = "EnumDescriptor"
 
+    const val PLUGIN_UTILS_FILE = "PluginUtils"
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/resolve/NamingConventions.kt:71]: Sorry for late notice, but that just came into my mind: by calling this file 'utils', we risk ending up with long file with a lot of unrelated stuff. I'd suggest avoid such technique and name files using logical groups, e.g. 'PluginExceptions'

**Review by sandwwraith — CHANGES_REQUESTED** [review]
LGTM, please fix comments

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt [hunk 6]
```diff
@@ -116,6 +148,153 @@ class SerializableIrGenerator(
...
+            )
+        } else {
+            irGetField(null, generateStaticDescriptorField())
+        }
+    }
+
+    private fun IrBlockBodyBuilder.generateStaticDescriptorField(): IrField {
+        val serialDescItType = serialDescClass.defaultType.toIrType()
+
+        val function = irClass.createInlinedFunction(
```
Comment (sandwwraith) [inline plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt:240]: Why there is a need in inlined function?

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt [hunk 7]
```diff
@@ -8,32 +8,53 @@ package org.jetbrains.kotlinx.serialization.compiler.backend.ir
...
 import org.jetbrains.kotlinx.serialization.compiler.extensions.SerializationPluginContext
 import org.jetbrains.kotlinx.serialization.compiler.resolve.*
+import org.jetbrains.kotlinx.serialization.compiler.resolve.SerialEntityNames.ARRAY_MASK_FIELD_MISSING_FUNC_FQ
 import org.jetbrains.kotlinx.serialization.compiler.resolve.SerialEntityNames.MISSING_FIELD_EXC
+import org.jetbrains.kotlinx.serialization.compiler.resolve.SerialEntityNames.SERIAL_DESC_FIELD
+import org.jetbrains.kotlinx.serialization.compiler.resolve.SerialEntityNames.SINGLE_MASK_FIELD_MISSING_FUNC_FQ
+import org.jetbrains.kotlinx.serialization.compiler.resolve.SerialEntityNames.initializedDescriptorFieldName
+
+// Another synthetic origin... it's time to stop?
+object SERIALIZABLE_PLUGIN_SYNTHETIC_ORIGIN : IrDeclarationOriginImpl("SERIALIZER SYNTHETIC", true)
```
[the remark this answers is not in the document]
Comment (shanshin) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt:38]: Existed origin have isSynthetic property = false and its bad idea create new synthetic fields and functions with this value.

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt (hunk 6 above)
Comment (shanshin) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt:240]: 1. I think it's good not to clutter up metadata of result class
2. I not found other good way to create clinit block on java platform - it's a good way to init field in <clinit>

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt [hunk 8]
```diff
@@ -108,6 +133,19 @@ interface IrBuilderExtension {
...
 
+    fun IrBuilderWithScope.createPrimitiveArrayOfExpression(
+        elementPrimitiveType: IrType,
+        arrayElements: List<IrExpression>
+    ): IrExpression {
+        val arrayType = compilerContext.irBuiltIns.primitiveArrayForType.getValue(elementPrimitiveType).defaultType
+        val arg0 = IrVarargImpl(startOffset, endOffset, arrayType, elementPrimitiveType, arrayElements)
+        val typeArguments = listOf(elementPrimitiveType)
+
+        return irCall(compilerContext.symbols.arrayOf, arrayType, typeArguments = typeArguments).apply {
```
[the remark this answers is not in the document]
Comment (shanshin) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/GeneratorHelpers.kt:144]: Yes, IR compiles to java like this
  int[] var44 = new int[]{seen1, seen2};

I didn't find symbol  intArrayOf for IrFunction, but now I rewrote it to compilerContext.irBuiltIns.primitiveArrayForType

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (sandwwraith) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt:38]: Judging by the look of `isSynthetic` usage I think we can just change `isSynthetic` to true for current `SERIALIZABLE_PLUGIN_ORIGIN`.

**Review by sandwwraith — APPROVED** [review]

File: plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt (hunk 7 above)
[the remark this answers is not in the document]
Comment (shanshin) [reply plugins/kotlin-serialization/kotlin-serialization-compiler/src/org/jetbrains/kotlinx/serialization/compiler/backend/ir/SerializableIrGenerator.kt]: Changed
